// inheretic/src/index.js
module.exports = (parent, packageName = 'package.json') => {
    const fs = require('fs');
    const path = require('path');
    function debug(msg) {
        console.log(msg());
    }
    function findPackages(folder, packages = []){
        const files = fs.readdirSync(folder);
        files.forEach((file) => {
            try {
                const resolved = path.join(folder, file);
                const stats = fs.lstatSync(resolved);
                if(stats.isDirectory()){
                    findPackages(resolved, packages);
                } 
                if (file.toLowerCase().endsWith(packageName)) {
                    packages.push(resolved);
                }
            } catch (x) {} // ignore - lstat errors for missing files everywhere
        });
        return packages;
    };
    parent = path.join(__dirname + '/../', parent);
    const packages = findPackages(parent);
    // TODO get this out of here
    if(!Array.prototype.flatMap){
        Array.prototype.flatMap = function(lambda) { 
            return Array.prototype.concat.apply([], this.map(lambda)); 
        };
    }
    const families = packages.flatMap((package) => {
        const parsed = JSON.parse(fs.readFileSync(package));
        parsed._filename = package;
        if(parsed.parent){
            const parent = path.resolve(path.dirname(package), parsed.parent);
            if(fs.existsSync(parent)){
                const json = JSON.parse(fs.readFileSync(parent));
                json._filename = parent;
                return[{
                    child: parsed, 
                    parent: json, 
                    toString: function(){return this.child.toString() + ":\n" + this.parent.toString() + "\n";}
                }];
            }
        } 
        return [];
    }).reduce((obj, family) => {
        obj[family.child._filename] = family;
        return obj;
    }, {});
    function inherit(parent, child){
        if(!child._generated){
            Object.keys(parent).forEach((key) => {
                const childValue = child[key];
                if(!childValue){
                    child[key] = parent[key];
                }
            });
            const _filename = child._filename;
            delete child._filename;
            delete child._generated;
            fs.writeFileSync(_filename.replace(packageName, "package.json"), JSON.stringify(child, null, 4));
            child._generated = true;
        } 
    }
    function chainInherit([key, ...tail] = Object.keys(families)){
        if(!key && tail.length === 0) return;
        const family = families[key];
        const parentFamily = families[family.parent._filename];
        if(parentFamily){
            const index = tail.indexOf(family.parent._filename);
            if(index > -1){
                tail.splice(index, 1);
            }
            chainInherit([family.parent._filename]);
            inherit(parentFamily.child, family.parent);
        }
        inherit(family.parent, family.child);
        chainInherit(tail);
    }
    chainInherit();
};
